plugins/overwatch.rb

```diff
@@ -0,0 +1,660 @@
+              session_log(sid, "Access error: #{self.state[sid][:error_count].to_s} of 3. Possibly migrating or routing not available.") # Only log this error three times
+              session_log(sid, "No access to routing, skipping session.") if self.state[sid][:error_count] >= 3 # Skip this session after it errored three times
+            else
+              session_log(sid, "Triggered an exception: #{e.class} <> #{e} #{e.backtrace}") # Log any other errors.
+            end
+          end
+        end
+        sleep(0.5) # This sleep time seams to work the best to keep processes from stepping on each other.
+      end
+    end
+
+    # Manages the top level of processing the sessions.
+    # Sessions must be fully operational before the subprocesses can work.
+    #
+    # @sid [int class] Session ID of the current session being processed
+    #
+    # @return [void] A useful return value is not expected here
+    def process(sid)
+      return unless framework.sessions[sid].info #Make sure session is fully active before processing
+      self.state[sid] ||= {}
+      store_session_info(sid)
+      process_routing(sid)
+      process_sessions(sid)
+    end
+
+    # Initialization and update session information.
+    #
+    # @sid [int class] Session ID of the current session
+    #
+    # @return [void] A useful return value is not expected here
+    def store_session_info(sid)
+      self.state[sid][:last_update] = Time.now.to_f
+      return if self.state[sid][:initialized]
+      self.state[sid][:info] = framework.sessions[sid].inspect
+      session_log(sid, "registered")
+      self.state[sid][:initialized] = true
+      self.state[sid][:error_count] = 0
+    end
+
+    # Prints lines of the log along with session information.
+    # Only prints to log if logging is enabled.
+    #
+    # @sid [int class] Session ID of the current session
+    # @msg [string class] Message to be displayed in the log
+    #
+    # @return [void] A useful return value is not expected here
+    def session_log(sid, msg)
+      return unless self.config[:log]
+      ::File.open(::File.join(self.config[:base], "session.log"), "a") do |fd|
+        fd.puts "#{Time.now.strftime('%Y-%m-%d %H:%M:%S')} Session #{sid} [#{self.state[sid][:info]}] #{msg}"
+      end
+    end
+
+    ##################################################################################
+    # Route Management Functions
+    ##################################################################################
+
+    # Perform the routing operations
+    # Response_timeout is changed to the defined stale time to stay consistant
+    # and not wait 300 seconds for timeout.
+    #
+    # @sid [int class] Session ID of of the session currently being processed.
+    #
+    # @return [int array] Array of session id's for stale sessions based on routing settings.
+    def process_routing(sid)
+      return unless self.config[:autoroute]
+      return if self.state[sid][:error_count] >= 3
```

Comment:
Direct manipulation of data structures requires external state management functions which can become convoluted with concurrency. Might be worth looking into converting the "reference nested hashes" approach to an object model encapsulating state management.